ardere/aws.py

```diff
@@ -0,0 +1,300 @@
+                {
+                    "Name": "tag:ECSCluster",
+                    "Values": [self._ecs_name]
+                }
+            ]
+        )
+        for page in response_iterator:
+            for reservation in page["Reservations"]:
+                for instance in reservation["Instances"]:
+                    # Determine if the instance is pending/running and count
+                    # 0 = Pending, 16 = Running, > is all shutting down, etc.
+                    if instance["State"]["Code"] <= 16:
+                        instance_dict[instance["InstanceType"]] += 1
+        return instance_dict
+
+    def calculate_missing_instances(self, desired, current):
+        # type: (Dict[str, int], Dict[str, int]) -> Dict[str, int]
+        """Determine how many of what instance types are needed to ensure
+        the current instance dict has all the desired instance count/types."""
+        needed = {}
+        for instance_type, instance_count in desired.items():
+            cur = current.get(instance_type, 0)
+            if cur < instance_count:
+                needed[instance_type] = instance_count - cur
+        return needed
+
+    def request_instances(self, instances):
+        # type: (Dict[str, int]) -> None
+        """Create requested types/quantities of instances for this cluster"""
+        ami_id = self.ecs_ami_ids["us-east-1"]
+        request_instances = []
+        for instance_type, instance_count in instances.items():
+            result = self._ec2_client.run_instances(
+                ImageId=ami_id,
+                KeyName="loads",
+                MinCount=instance_count,
+                MaxCount=instance_count,
+                InstanceType=instance_type,
+                UserData="#!/bin/bash \necho ECS_CLUSTER='" + self._ecs_name +
+                         "' >> /etc/ecs/ecs.config",
+                IamInstanceProfile={"Arn": os.environ["ecs_profile"]}
+            )
+
+            # Track returned instances for tagging step
+            request_instances.extend([x["InstanceId"] for x in
+                                      result["Instances"]])
+
+        self._ec2_client.create_tags(
+            Resources=request_instances,
+            Tags=[
+                dict(Key="Owner", Value="ardere"),
+                dict(Key="ECSCluster", Value=self._ecs_name)
+            ]
+        )
+
+    def create_service(self, step):
+        # type: (Dict[str, Any]) -> Dict[str, Any]
+        """Creates an ECS service for a step and returns its info"""
+        logger.info("CreateService called with: {}".format(step))
+
+        # Prep the shell command
+        shell_command = [
+            'sh', '-c', '"$WAITFORCLUSTER"',
+            'waitforcluster.sh', self.s3_ready_file,
+            str(step.get("run_delay", 0))
+        ]
+        shell_command2 = ' '.join(shell_command) + ' && ' + step[
+            "additional_command_args"]
+        shell_command3 = ['sh', '-c', '{}'.format(shell_command2)]
```

Comment:
It's mainly related to how the sh is nesting a shell command with quotes (and docker wackiness).